### PR TITLE
fix: restore active markets from checkpoint proposals

### LINKED ISSUE #5279: Incorrect proposal handling with checkpoints
Checkpoints do not behave correctly when reloading  markets which haven't settle.
Per spec markets should be restarted if they were not closed and settle and the opening auction should start again.

https://github.com/vegaprotocol/specs-internal/blob/master/non-protocol-specs/0005-NP-LIMN-limited_network_life.md#restoring-a-checkpoint

### PR DESCRIPTION
**Pull request by ze97286**
Closes #5348 #5279

### CODE REVIEW

File: assets/assets.go [hunk 1]
```diff
@@ -58,13 +58,16 @@ type Service struct {
 	isValidator bool
 }
 
+type TimeService interface {
+	GetTimeNow() time.Time
+}
+
```
Comment (jeremyletang) [inline assets/assets.go:64]: I don't think we need this interface?

File: execution/engine.go [hunk 2]
```diff
@@ -280,8 +280,29 @@ func (e *Engine) SubmitMarketWithLiquidityProvision(ctx context.Context, marketC
...
+	return e.submitOrRestoreMarketWithLiquidityProvision(ctx, marketConfig, lp, party, lpID, deterministicId, true)
+}
+
+// RestoreMarketWithLiquidityProvision is very similar to SubmitMarketWithLiquidityProvision without updating the market tracker as it is restored separately from checkpoint.
+func (e *Engine) RestoreMarketWithLiquidityProvision(ctx context.Context, marketConfig *types.Market, lp *types.LiquidityProvisionSubmission, lpID, deterministicId string) error {
+	proposer := e.marketActivityTracker.GetProposer(marketConfig.ID)
+	if len(proposer) == 0 {
+		return ErrMarketDoesNotExist
+	}
+	return e.submitOrRestoreMarketWithLiquidityProvision(ctx, marketConfig, lp, proposer, lpID, deterministicId, false)
```
Comment (jeremyletang) [inline execution/engine.go:292]: Will it be fine here to use the market ID as the proposer of the market? will we not be submitting the LP for a wrong party ID?

Comment (ze97286) [reply execution/engine.go:292]: we're not submitting the market ID, we're submitting the proposer of the market as the proposer.

Comment (jeremyletang) [reply execution/engine.go:292]: oh right. I'm an idot indeed.

File: governance/engine.go [hunk 3]
```diff
@@ -495,7 +498,7 @@ func (e *Engine) intoToSubmit(ctx context.Context, p *types.Proposal) (*ToSubmit
 		enactTime := time.Unix(p.Terms.EnactmentTimestamp, 0)
 		newMarket := p.Terms.GetNewMarket()
 
-		auctionDuration := enactTime.Sub(closeTime)
+		auctionDuration := time.Duration((enactTime.Unix() - closeTime.Unix()) * int64(time.Second))
```
Comment (wwestgarth) [inline governance/engine.go:501]: Minor: I'm sure theres a good reason you're stripping off all the precision but given that `enactTime` is already just in precision seconds, I think all you need to do is here is`closeTime := e.timeService.GetTimeNow().Truncate(time.Second)`

Also that `FIXME` comment about the auctionState, do we not also need to trunctate the time passed into `NewAuctionState`?

File: governance/checkpoint.go [hunk 4]
```diff
@@ -31,31 +37,73 @@ func (e *Engine) Checkpoint() ([]byte, error) {
...
 			return err
 		}
+
+		switch prop.Terms.Change.GetTermType() {
+		case types.ProposalTermsTypeNewMarket:
+			// if the proposal is for a new market we want to restore it such that it will be in opening auction
+			if p.Terms.EnactmentTimestamp <= now.Unix() {
+				prop.Terms.EnactmentTimestamp = now.Add(duration).Unix()
+			}
+			if toSubmit, err := e.intoToSubmit(ctx, prop); err == nil {
```
Comment (wwestgarth) [inline governance/checkpoint.go:80]: If `intoToSubmit` fails we're just silently ignoring it, is that intentional?

Comment (ze97286) [reply governance/checkpoint.go:80]: I think it pretty much can't really fail (famous last words) but I'll add panic.

File: governance/engine.go (hunk 3 above)
Comment (ze97286) [reply governance/engine.go]: addressed

File: assets/assets.go (hunk 1 above)
Comment (ze97286) [reply assets/assets.go]: removed

**Review by wwestgarth — DISMISSED** [review]

**Review by wwestgarth — APPROVED** [review]

**Pull request merged**